rust/metering/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+                }
+            }
+        }
+
+        /// A blanket-impl of the `submit` method for all metering events.
+        #[async_trait::async_trait]
+        impl<T> SubmitExt for T
+        where
+            T: MeteringEvent + Send + 'static,
+        {
+            async fn submit(self) {
+                let maybe_current_span = Some(tracing::Span::current());
+                let receiver: Box<
+                    dyn chroma_system::ReceiverForMessage<Box<dyn MeteringEvent>>,
+                > = {
+                    let lock = RECEIVER.lock();
+                    (*lock).clone_box()
+                };
+                let boxed_metering_event: Box<dyn MeteringEvent> = Box::new(self);
+                if let Err(error) = receiver.send(boxed_metering_event, maybe_current_span).await {
+                    tracing::error!("Unable to send meter event: {error}");
+                }
+            }
+        }
+
+        /// A global map stores a map of thread IDs two stacks of metering events.
+        static EVENT_STACKS: once_cell::sync::Lazy<
+            parking_lot::Mutex<std::collections::HashMap<std::thread::ThreadId, Vec<(std::any::TypeId, Box<dyn MeteringEvent>)>>>> =
+            once_cell::sync::Lazy::new(|| parking_lot::Mutex::new(std::collections::HashMap::new()));
+
+        /// Gets a mutable reference to the current thread's event stack.
+        fn get_stack_for_current_thread<'a>(
+            event_stacks: &'a mut std::collections::HashMap<std::thread::ThreadId, Vec<(std::any::TypeId, Box<dyn MeteringEvent>)>>,
+        ) -> &'a mut Vec<(std::any::TypeId, Box<dyn MeteringEvent>)> {
+            let thread_id = std::thread::current().id();
+            event_stacks.entry(thread_id).or_default()
+        }
+
+        /// Pops the top entry (TypeId, Box<dyn MeteringEvent>) from the current thread and returns it.
+        /// If no event exists, returns `None`.
+        fn pop_from_thread(thread_id: std::thread::ThreadId) -> Option<(std::any::TypeId, Box<dyn MeteringEvent>)> {
+            let mut event_stacks = EVENT_STACKS.lock();
+            if let Some(event_stack) = event_stacks.get_mut(&thread_id) {
+                event_stack.pop()
+            } else {
+                None
+            }
+        }
+
+        /// Pushes an entry (TypeId, Box<dyn MeteringEvent>) onto the current thread's stack.
+        fn push_to_thread(thread_id: std::thread::ThreadId, entry: (std::any::TypeId, Box<dyn MeteringEvent>)) {
+            let mut event_stacks = EVENT_STACKS.lock();
+            let event_stack = event_stacks.entry(thread_id).or_default();
+            event_stack.push(entry);
+        }
+
+        /// Peek at the top of the current thread’s stack, returning a raw pointer if an event is present;
+        /// else returns `None`.  If `None`, callers should use a `BlankMeteringEvent`
+        fn peek_current_thread_raw() -> Option<*mut dyn MeteringEvent> {
```

Comment:
Why are we using raw pointers, which are unsafe?

https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/raw-pointers.html#:~:text=Rust%20has%20a%20number%20of,and%20are%20unsafe%20to%20use.